apps/ejabberd/src/mongoose_deprecations.erl

```diff
@@ -0,0 +1,101 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% File    : mongoose_deprecations.erl
+%% Author  : Dominik Stanaszek <dominik.stanaszek@erlang-solutions.com>
+%% Purpose : More generic deprecation handling
+%% Created : 10 Oct 2017
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc This module is responsible for initialising
+%% and stopping stuff needed to handle different deprecation
+%% warinings as well as for exposing API for loging these
+%% deprecations.
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+-module(mongoose_deprecations).
+-author("dominik.stanaszek@erlang-solutions.com").
+
+-export([start/0, stop/0, log/1, log/2]).
+
+-include("ejabberd.hrl").
+
+-type deprecation() :: atom().
+-type log_level() :: warning | error.
+
+-spec start() -> ok.
+start() ->
+    start_mam02_reminder(),
+    ok.
+
+-spec stop() -> ok.
+stop() ->
+    stop_mam02_reminder(),
+    ok.
+
+%% @doc Performs a log specific for a deprecation
+%% which is passed as an atom.
+%% It can also log a custom string.
+%% Logs as ERROR message.
+-spec log(deprecation() | string()) -> ok.
+log(mam02_archived) ->
+    maybe_log_mamv02_deprecation_error();
```

Comment:
This is not very generic. This module shouldn't provide an implementation specific for certain deprecations, as it unnecessarily increases coupling. The API should be `log(Tag :: atom(), Level :: atom(), FormatStr :: string(), Params :: list())` (with `log(Tag, FormatStr, Params)` may still default to `error` log level). There shouldn't be per-deprecation table but one global for all deprecations and its elements' keys should be `{Tag, last_logged}` or just `Tag`. I think we may stick to common cooldown time for all deprecations for now.

Reply:
I thought of this module as tightly coupled with those that want to "call themselves" deprecated, as a module that knows about specific modules and the way each of them wants to annnounce the deprecation. So that now module itself just  needs to trigger `log/1` function whenever it wants to log deprecation message and here is the logic responsible for cooldowns, log format, preparing necessary ETS tables etc.

What would be the gain for creating a module with only a log message?
Moreover - it wouldn't be too generic if we say each deprecation has some kind of cooldown. Maybe some deprecation messages are shown at Mongoose's start or shutdown and not triggered by some event as `<archived/>` is?

I see this module as one that:
- exposes API for starting and stoping it (where it creates all the stuff that is needed to handle different types of deprecations e.g. creating ETS tables or registering hooks
- exposes API to log such a warning

@fenek, tell me how is your vision different from this one so that I can make necessary corrections. :)

Reply:
We could make it the other way round:
make mongoose_deprecations' functions callbacks and make modules implement them. That would loosen the coupling between them and this new module but on the other hand we will have this quite a huge piece of code that is only responsible for deprecating (which is not part of the main job of the module).

Reply:
We can also go in the direction that all deprecations look alike - they have some cooldown and if module calls our `log` function we first check if it's not logged too early.

Reply:
So, as I guess you probably meant the third options, I will stick to that